diff --git a/llmc/layernorm.cuh b/llmc/layernorm.cuh
--- a/llmc/layernorm.cuh
+++ b/llmc/layernorm.cuh
@@ -165,7 +165,7 @@ __global__ void fused_residual_forward_kernel5(floatX* residual, floatX* normed, flo
     __syncthreads();
 
     int idx = blockIdx.x * blockDim.y + threadIdx.y;
-    if(idx > N) return;
+    if(idx >= N) return;
 
     // adjust pointers to current token
     residual += C * idx;
